OpenRA.Mods.Common/Traits/Cargo.cs

```diff
@@ -201,11 +217,32 @@ public string VoicePhraseForOrder(Actor self, Order order)
 		public bool HasSpace(int weight) { return totalWeight + reservedWeight + weight <= Info.MaxWeight; }
 		public bool IsEmpty(Actor self) { return cargo.Count == 0; }
 
-		public Actor Peek(Actor self) { return cargo.Peek(); }
+		public Actor GetLastEntered(Actor self)
+		{
+			if (IsEmpty(self))
+				throw new InvalidOperationException("GetLastEntered called on an empty transport ({0}).".F(self));
+
+			return cargo[cargo.Count - 1];
+		}
 
-		public Actor Unload(Actor self)
+		public Actor UnloadLastEntered(Actor self)
 		{
-			var a = cargo.Pop();
+			if (IsEmpty(self))
+				throw new InvalidOperationException("UnloadLastEntered called on an empty transport ({0}).".F(self));
+
+			var a = cargo[cargo.Count - 1];
+			return UnloadSpecific(self, a);
+		}
+
+		public Actor UnloadSpecific(Actor self, Actor toUnload)
+		{
+			var idx = cargo.IndexOf(toUnload);
+			if (idx == -1)
+				throw new InvalidOperationException("Cargo for {0} does not contain actor {1} but tried to unload it."
```

Comment:
I'm not sure if I like `Cargo` throwing exceptions left and right. Why not just return null and make the caller check the return value? The caller needs a check with your approach already (IsEmpty), so turning that check into a null check (if the caller is even interested in it) and thus simplifying `Cargo` would be the better solution IMO.